### PR TITLE
Add macro support in jump to definition feature

### LINKED ISSUE #91174: Jump to definition doesn't generate a link for macro calls
On [this page](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_errors/diagnostic_builder.rs.html), the macros calls (`forward!`, `concat!`, ...) are not clickable to jump to their definitions.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #91174.

To do so, I check if the span comes from an expansion, and if so, I infer the original macro `DefId` or `Span` depending if it's a defined in the current crate or not.

There is one limitation due to macro expansion though:

```rust
macro_rules! yolo { () => {}}

fn foo() {
    yolo!();
}
```

In `foo`, `yolo!` won't be linked because after expansion, it is replaced by nothing (which seems logical). So I can't get an item from the `Visitor` from which I could tell if its `Span` comes from an expansion.

I added a test for this specific limitation alongside others.

Demo: https://rustdoc.crud.net/imperio/macro-jump-to-def/src/foo/check-source-code-urls-to-def-std.rs.html

As for the empty macro issue that cannot create a jump to definition, you can see it [here](https://rustdoc.crud.net/imperio/macro-jump-to-def/src/foo/check-source-code-urls-to-def-std.rs.html#35).

Part of https://github.com/rust-lang/rust/issues/89095.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/highlight.rs [hunk 1]
```diff
@@ -171,8 +177,22 @@ impl Class {
...
+            | Self::Op
+            | Self::MacroNonTerminal
+            | Self::String
+            | Self::Number
+            | Self::Bool
+            | Self::Lifetime
+            | Self::PreludeTy
+            | Self::PreludeVal
+            | Self::QuestionMark
+            | Self::Decoration(_) => None,
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/highlight.rs:195]: I made this change in this PR because, thanks to it, I lost 20 minutes of my life trying to understand why the `Class::Macro` wasn't getting its `span` checked against the "span map". If you prefer this change to be in another PR, please tell and I'll send another one and remove it from this PR.

File: src/librustdoc/html/highlight.rs [hunk 2]
```diff
@@ -748,11 +796,51 @@ fn string<T: Display>(
...
         }
     }
-    write!(out, "<span class=\"{}\">{}</span>", klass.as_html(), text_s);
+    write!(out, "<span class=\"{}\">{}", klass.as_html(), text_s);
+    Some("</span>")
+}
+
+/// This function is to get the external macro path because they are not in the cache used n
+/// `href_with_root_path`.
+fn generate_macro_def_id_path(context_info: &ContextInfo<'_, '_, '_>, def_id: DefId) -> String {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/highlight.rs:809]: I'm not super happy about this part: the `cache` contains external items but not macros, meaning that I had to compute it here. I originally thought about adding the macro paths into the `cache` but after a discussion with @jyn514, we thought that maybe it would be worth it to compute macros' paths on demand.

To be noted: this is a mix between `clean::inline::record_extern_fqn` and `html::format::href_with_root_path`. Does it look reasonable as is?

File: src/librustdoc/html/render/span_map.rs [hunk 3]
```diff
@@ -62,32 +63,59 @@ struct SpanMapVisitor<'tcx> {
...
+            self.matches.insert(path.span, LinkFromSrc::Local(clean::Span::new(span)));
         } else if let Some(def_id) = info {
-            self.matches.insert(path_span.unwrap_or(path.span), LinkFromSrc::External(def_id));
+            self.matches.insert(path.span, LinkFromSrc::External(def_id));
+        }
+    }
+
+    /// Adds the macro call into the span map. Returns `true` if the `span` was inside a macro
+    /// expansion, whether or not it was added to the span map.
+    fn handle_macro(&mut self, span: Span) -> bool {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/span_map.rs:90]: The idea for the macro support is to check if the current `Span` comes from expansion. If so, we loop until we find the macro definition by using `outer_expn_data` in a loop. Finally, we get the information about the macro itself (`span` if "local", `DefId` otherwise) and store it inside the span map.

File: src/test/rustdoc/check-source-code-urls-to-def-std.rs [hunk 4]
```diff
@@ -15,3 +15,28 @@ pub fn foo(a: u32, b: &str, c: String) {
...
+}
+
+macro_rules! data{
+    ($x:expr) => { $x * 2 }
+}
+
+pub fn another_foo() {
+    // This is known limitation: if the macro doesn't generate anything, the visitor
+    // can't find any item or anything that could tell us that it comes from expansion.
+    // @!has - '//a[@href="../../src/foo/check-source-code-urls-to-def-std.rs.html#19"]' 'yolo!'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/check-source-code-urls-to-def-std.rs:34]: A way to go around that limitation would be to `resolve` the macro (just like in intra-doc links). Might be better in a follow-up PR though.

File: src/test/rustdoc/check-source-code-urls-to-def-std.rs [hunk 5]
```diff
@@ -15,3 +15,28 @@ pub fn foo(a: u32, b: &str, c: String) {
...
+
+macro_rules! yolo { () => {}}
+
+fn bar(a: i32) {}
+
+macro_rules! bar {
+    ($a:ident) => { bar($a) }
+}
+
+macro_rules! data{
```
Comment (camelid) [inline src/test/rustdoc/check-source-code-urls-to-def-std.rs:27]: nit:

```suggestion
macro_rules! data {
```

Comment (GuillaumeGomez) [reply src/test/rustdoc/check-source-code-urls-to-def-std.rs]: Thanks, fixed!

File: src/test/rustdoc/check-source-code-urls-to-def-std.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc/check-source-code-urls-to-def-std.rs:34]: After taking a deeper look, it would require the `Resolver` from `DocContext`, so might be a bit bigger than what I originally imagined. I think doing in another PR makes even more sense.

Comment (jyn514) [reply src/test/rustdoc/check-source-code-urls-to-def-std.rs:34]: Please don't add more string-based resolving ... I haven't thought about this problem in detail, but looking it up by name alone doesn't seem like the right approach.

Comment (GuillaumeGomez) [reply src/test/rustdoc/check-source-code-urls-to-def-std.rs:34]: Do you prefer a rustdoc-gui test instead?

**Comment by jyn514** [comment]
I won't have a chance to review this any time soon.

r? @camelid

**Comment by camelid** [comment]
I don't have knowledge of this code. r? rust-lang/rustdoc

**Comment by Dylan-DPC** [comment]
@jsha this is waiting for your review

**Comment by GuillaumeGomez** [comment]
I'll fix the conflict once https://github.com/rust-lang/rust/pull/96636 is merged.

**Comment by jsha** [comment]
#96636 is now merged.

**Comment by GuillaumeGomez** [comment]
Here you go: https://rustdoc.crud.net/imperio/macro-jump-to-def/src/foo/check-source-code-urls-to-def-std.rs.html

As for the empty macro issue that cannot create a jump to definition, you can see it [here](https://rustdoc.crud.net/imperio/macro-jump-to-def/src/foo/check-source-code-urls-to-def-std.rs.html#35).

**Comment by Kobzol** [comment]
I wonder why `eprintln!` in the demo links to doc and not source?

**Comment by GuillaumeGomez** [comment]
It's because the code tried to find it location based on its `Span` and failed, using then the documentation page (which we can generate pretty easily).

**Comment by GuillaumeGomez** [comment]
Fixed the conflict.

**Comment by GuillaumeGomez** [comment]
Completely anecdoctical: since it'll conflict with https://github.com/rust-lang/rust/pull/89157, I have a branch with the changes ready to fix the conflicts [here](https://github.com/rust-lang/rust/compare/master...GuillaumeGomez:jump-to-def-extended?expand=1).

File: src/librustdoc/html/render/span_map.rs [hunk 6]
```diff
@@ -63,32 +64,59 @@ struct SpanMapVisitor<'tcx> {
...
+        if span.from_expansion() {
+            let mut data = span.ctxt().outer_expn_data();
+            let mut call_site = data.call_site;
+            while call_site.from_expansion() {
```
Comment (jsha) [inline src/librustdoc/html/render/span_map.rs:95]: Is the idea here that if a macro invokes a macro, it will result in a code span that is the result of an expansion, which is the result of another expansion; and the macro name that users see in source view is always the one that caused the outermost expansion?

I think it would be useful to add a comment explaining that.

File: src/librustdoc/html/render/span_map.rs (hunk 3 above)
Comment (jsha) [reply src/librustdoc/html/render/span_map.rs:91]: This is useful to know, and should be written in a comment.

File: src/librustdoc/html/render/span_map.rs [hunk 7]
```diff
@@ -63,32 +64,59 @@ struct SpanMapVisitor<'tcx> {
...
+                let new_span = data.call_site;
+                let macro_name = macro_name.as_str();
+                // The "call_site" includes the whole macro with its "arguments". We only want
+                // the macro name.
+                let new_span = new_span.with_hi(new_span.lo() + BytePos(macro_name.len() as u32));
+                self.matches.insert(new_span, link_from_src);
+            }
+            true
+        } else {
+            false
```
Comment (jsha) [inline src/librustdoc/html/render/span_map.rs:119]: The nested `if`s and `if let`s here get a little intense and hard to follow. Is there a way to restructure it? For instance you could arrange an early return in the case where it's not a bang macro:

```
let macro_name = match data.kind {
  ExpnKind::Macro(MacroKind::Bang, macro_name) => macro_name,
  _ => return false
};
```

And similarly an early return for when the def_id is not available:

```
let link_from_src = match data.macro_def_id {
  Some(did) if did.is_local() => LinkFromSrc::Local...,
  Some(did) => LinkFromSrc::External...,
  None => return true,
}
```

For that matter, while we're talking early returns - `if span.expansion()` up above should have its sense inverted and immediately return if false.

File: src/librustdoc/html/render/span_map.rs [hunk 8]
```diff
@@ -143,12 +174,18 @@ impl<'tcx> Visitor<'tcx> for SpanMapVisitor<'tcx> {
                     );
                 }
             }
+        } else if self.handle_macro(expr.span) {
+            // We don't want to deeper into the macro.
```
Comment (jsha) [inline src/librustdoc/html/render/span_map.rs:178]:
```suggestion
            // We don't want to go deeper into the macro.
```

File: src/librustdoc/html/render/span_map.rs [hunk 9]
```diff
@@ -63,32 +64,59 @@ struct SpanMapVisitor<'tcx> {
 
 impl<'tcx> SpanMapVisitor<'tcx> {
     /// This function is where we handle `hir::Path` elements and add them into the "span map".
-    fn handle_path(&mut self, path: &rustc_hir::Path<'_>, path_span: Option<Span>) {
+    fn handle_path(&mut self, path: &rustc_hir::Path<'_>) {
```
Comment (jsha) [inline src/librustdoc/html/render/span_map.rs:67]: Why does handle_path not need path_span anymore?

**Review by jsha — COMMENTED** [review]
Some initial thoughts; I'll take a second pass soon. Thanks for your patience on this review. :-)

File: src/librustdoc/html/highlight.rs [hunk 10]
```diff
@@ -658,13 +677,18 @@ impl<'a> Classifier<'a> {
 
 /// Called when we start processing a span of text that should be highlighted.
 /// The `Class` argument specifies how it should be highlighted.
-fn enter_span(out: &mut Buffer, klass: Class) {
-    write!(out, "<span class=\"{}\">", klass.as_html());
+fn enter_span(
+    out: &mut Buffer,
+    klass: Class,
+    context_info: &Option<ContextInfo<'_, '_, '_>>,
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:683]: Sure, this is a good idea!

File: src/librustdoc/html/render/span_map.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:95]: I'll put the explanation in the comment directly then please tell me if it's clear enough.

File: src/librustdoc/html/render/span_map.rs (hunk 9 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:67]: The last argument was always `None` so this is basically a cleanup that was forgotten in a previous update.

File: src/librustdoc/html/render/span_map.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:91]: I'll update this comment then.

File: src/librustdoc/html/render/span_map.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:119]: It indeed makes the code much simpler to read. Thanks for the suggestion!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
> Some initial thoughts; I'll take a second pass soon. Thanks for your patience on this review. :-)

No problem, thanks a lot for going through this! Don't hesitate if you need explanations on something.

File: src/librustdoc/html/highlight.rs [hunk 11]
```diff
@@ -727,27 +766,76 @@ fn string<T: Display>(
...
+    };
+
+    let last = path.pop().unwrap();
+    let last = format!("macro.{}.html", last);
+    if path.is_empty() {
+        format!("{}/{}", url_parts.join("/"), last)
+    } else {
+        format!("{}/{}/{}", url_parts.join("/"), path.join("/"), last)
+    }
 }
```
Comment (jsha) [inline src/librustdoc/html/highlight.rs:839]: I had a bunch of comments on this function so I figured I'd just try rewriting it locally to my taste. Let me know what you think!

```suggestion
/// Return the URL for a macro at the given DefId.
/// 
/// This needs its own function because macros are not in the cache used by `href_with_root_path`.
fn generate_macro_def_id_path(href_context: &HrefContext<'_, '_, '_>, def_id: DefId) -> String {
    let tcx = href_context.context.shared.tcx;
    let relative_path = tcx.def_path(def_id).data.into_iter().filter_map(|elem| {
        // extern blocks have an empty name
        let s = elem.data.to_string();
        if !s.is_empty() { Some(s) } else { None }
    });

    let crate_name = tcx.crate_name(def_id.krate).to_string();
    // In macros 2.0, macros can have a path. If available, find that path.
    // https://github.com/rust-lang/rfcs/blob/master/text/1584-macros.md
    let mut path = match CStore::from_tcx(tcx).load_macro_untracked(def_id, tcx.sess) {
        LoadedMacro::MacroDef(def, _)
            if matches!(&def.kind,
                ast::ItemKind::MacroDef(ast_def) if !ast_def.macro_rules) => {
            once(crate_name.clone()).chain(relative_path).collect()
        },

        _ => vec![crate_name.clone(), relative_path.last().expect("macro's path has no name")],
    };

    let macro_name = path.pop().expect("macro has empty path when linkifying");
    let root = match href_context.context.cache().extern_locations[&def_id.krate] {
        ExternalLocation::Remote(remote) => remote.clone(),
        ExternalLocation::Local => format!("{}{}", href_context.root_path, crate_name),
        ExternalLocation::Unknown => panic!("unknown crate"),
    };
    format!("{}{}/macro.{}.html", root, path.join("/"), macro_name)
}
```

File: src/librustdoc/html/highlight.rs [hunk 12]
```diff
@@ -687,15 +711,30 @@ fn string<T: Display>(
...
+        out.write_str(closing_tag);
+    }
+}
+
+fn string_without_closing_tag<T: Display>(
+    out: &mut Buffer,
+    text: T,
+    klass: Option<Class>,
+    href_context: &Option<HrefContext<'_, '_, '_>>,
+) -> Option<&'static str> {
```
Comment (jsha) [inline src/librustdoc/html/highlight.rs:726]: We also use the "optional close tag" pattern in `render_impl` in render/mod.rs.

I think we can make things a little clearer by using the type system:

```
enum ClosingTag {
  A,
  Span,
  Details,
  Div,
  None,
}

impl Display for ClosingTag {
 ...
}
```

Then this function could return `ClosingTag` and in render/mod.rs we could keep track of a `Vec<ClosingTag>`.

File: src/librustdoc/html/highlight.rs [hunk 13]
```diff
@@ -727,27 +766,76 @@ fn string<T: Display>(
...
-                        format::href_with_root_path(*def_id, context, Some(context_info.root_path))
-                            .ok()
+                        format::href_with_root_path(*def_id, context, Some(href_context.root_path))
                             .map(|(url, _, _)| url)
+                            .or_else(|e| {
+                                if e == format::HrefError::NotInExternalCache
+                                    && matches!(klass, Class::Macro(_))
+                                {
+                                    Ok(generate_macro_def_id_path(href_context, *def_id))
+                                } else {
```
Comment (jsha) [inline src/librustdoc/html/highlight.rs:778]: Why not incorporate this clause into `format::href_with_root_path` itself? It would require refactoring `href_with_root_path` to take an `HrefContext`, and that would require moving `HrefContext` into `format`, but I think those are both reasonable steps. It would be nice to have a more complete and accurate `href_with_root_path` rather than have a special case that only applies when linkifying.

**Review by jsha — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/highlight.rs [hunk 14]
```diff
@@ -44,7 +48,7 @@ pub(crate) fn render_with_highlighting(
     tooltip: Option<(Option<Edition>, &str)>,
     edition: Edition,
     extra_content: Option<Buffer>,
-    context_info: Option<ContextInfo<'_, '_, '_>>,
+    href_context: Option<HrefContext<'_, '_, '_>>,
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:51]: It's definitely worth an issue so it can be fixed. I'll let you open it with your (great!) explanations. :)

File: src/librustdoc/html/render/span_map.rs [hunk 15]
```diff
@@ -63,33 +64,72 @@ struct SpanMapVisitor<'tcx> {
...
+        // origin of this expansion.
+        while call_site.from_expansion() {
+            data = call_site.ctxt().outer_expn_data();
+            call_site = data.call_site;
         }
+
+        let macro_name = match data.kind {
+            ExpnKind::Macro(MacroKind::Bang, macro_name) => macro_name,
+            // Even though we don't handle this kind of macro, this `data` still comes from
+            // expansion so we return `true` so we don't any deeper in this code.
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:116]: I keep making this typo. ><

File: src/librustdoc/html/highlight.rs [hunk 16]
```diff
@@ -85,31 +89,32 @@ fn write_header(out: &mut Buffer, class: Option<&str>, extra_content: Option<Buf
...
         decoration_info,
     )
     .highlight(&mut |highlight| {
         match highlight {
-            Highlight::Token { text, class } => string(out, Escape(text), class, &context_info),
-            Highlight::EnterSpan { class } => enter_span(out, class),
-            Highlight::ExitSpan => exit_span(out),
+            Highlight::Token { text, class } => string(out, Escape(text), class, &href_context),
+            Highlight::EnterSpan { class } => closing_tag = enter_span(out, class, &href_context),
+            Highlight::ExitSpan => exit_span(out, &closing_tag),
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:117]: I'm not sure but for safety, I'll use a `Vec` so it can never happen.

File: src/librustdoc/html/highlight.rs [hunk 17]
```diff
@@ -658,13 +677,18 @@ impl<'a> Classifier<'a> {
...
 /// The `Class` argument specifies how it should be highlighted.
-fn enter_span(out: &mut Buffer, klass: Class) {
-    write!(out, "<span class=\"{}\">", klass.as_html());
+fn enter_span(
+    out: &mut Buffer,
+    klass: Class,
+    href_context: &Option<HrefContext<'_, '_, '_>>,
+) -> &'static str {
+    string_without_closing_tag(out, "", Some(klass), href_context)
+        .expect("no closing tag to close wrapper...")
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:686]: :+1:

File: src/librustdoc/html/highlight.rs [hunk 18]
```diff
@@ -687,15 +711,30 @@ fn string<T: Display>(
...
     klass: Option<Class>,
-    context_info: &Option<ContextInfo<'_, '_, '_>>,
+    href_context: &Option<HrefContext<'_, '_, '_>>,
 ) {
+    if let Some(closing_tag) = string_without_closing_tag(out, text, klass, href_context) {
+        out.write_str(closing_tag);
+    }
+}
+
+fn string_without_closing_tag<T: Display>(
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:721]: Good point, I'll add a doc comment and rename it.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:721]: I sopke too fast, for once I think it's a good name because it's based on the `string` function just above. I can rename both though if you have an idea for the name to use.

File: src/librustdoc/html/highlight.rs (hunk 12 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:726]: This function doesn't go "up" to `render/mod.rs`, it's just used in this file as it generates the whole content of the page/code block.

Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:726]: Sorry, I misunderstood your comment. I'm not sure if it's a good idea or not. Might be worth to discuss it "outside" of this PR maybe?

File: src/librustdoc/html/highlight.rs [hunk 19]
```diff
@@ -727,27 +766,76 @@ fn string<T: Display>(
...
+    write!(out, "<span class=\"{}\">{}", klass.as_html(), text_s);
+    Some("</span>")
+}
+
+/// This function is to get the external macro path because they are not in the cache used n
+/// `href_with_root_path`.
+fn generate_macro_def_id_path(href_context: &HrefContext<'_, '_, '_>, def_id: DefId) -> String {
+    let tcx = href_context.context.shared.tcx;
+    let crate_name = tcx.crate_name(def_id.krate).to_string();
+    let cache = &href_context.context.cache();
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:807]: Very good point!

File: src/librustdoc/html/highlight.rs [hunk 20]
```diff
@@ -727,27 +766,76 @@ fn string<T: Display>(
...
+        once(crate_name.clone()).chain(relative).collect()
+    } else {
+        vec![crate_name.clone(), relative.last().expect("relative was empty")]
+    };
+
+    let url_parts = match cache.extern_locations[&def_id.krate] {
+        ExternalLocation::Remote(ref s) => vec![s.trim_end_matches('/')],
+        ExternalLocation::Local => vec![href_context.root_path.trim_end_matches('/'), &crate_name],
+        ExternalLocation::Unknown => panic!("unknown crate"),
+    };
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:830]: I can actually make it even simpler combined with your next comment by merging both. :)

File: src/librustdoc/html/highlight.rs (hunk 11 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:839]: It's actually a bit similar to what I wrote except mine is a bit longer but I find it easier to read. You will tell me what you think of it. :)

File: src/librustdoc/html/render/span_map.rs [hunk 21]
```diff
@@ -63,33 +64,72 @@ struct SpanMapVisitor<'tcx> {
...
+                LinkFromSrc::Local(clean::Span::new(data.def_site))
+            }
+            Some(macro_def_id) => LinkFromSrc::External(macro_def_id),
+            None => return true,
+        };
+        let new_span = data.call_site;
+        let macro_name = macro_name.as_str();
+        // The "call_site" includes the whole macro with its "arguments". We only want
+        // the macro name.
+        let new_span = new_span.with_hi(new_span.lo() + BytePos(macro_name.len() as u32));
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:130]: It's not possible as is because we need `new_span.lo()` but I'll store it instead.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:130]: Hum actually it renders ugly too because I need to write `let new_span_lo = data.call_site.lo();`. Not sure what you think about it?

[the remark this answers is not in the document]
Comment (jsha) [reply src/librustdoc/html/render/span_map.rs:130]: This is fine - it was a small tweak and I didn't notice that we actually needed `macro_name` in two places.

File: src/librustdoc/html/highlight.rs (hunk 12 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:741]: The contexts and why it's needed are different so having a common mechanism seems a bit weird for me. Not saying it's a bad idea, just that I'm not really convinced it's needed to add a type to handle this.

File: src/librustdoc/html/highlight.rs (hunk 13 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:793]: `href_with_root_path` handles external items but the macros are not in the cache used in this function. Hence why I needed to add this function.

Comment (jsha) [reply src/librustdoc/html/highlight.rs:793]: So what I'm saying is:

 - `href_with_root_path` handles external items in general (functions, structs, etc).
 - `href_with_root_path` does not currently handle external macros.
 - Use cases outside of highlighting don't call `href_with_root_path` for external macros, but that's okay.
 - If `href_with_root_path` could handle external macros just as well as it handles functions and structs, it would be a more complete function.
 - That makes it easier to understand the code base in general.

Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:793]: That would force me to pass `Option<HrefContext>` as last argument because the function is called by `href` and `href_with_root_path` is already quite big. Are you sure you don't prefer me creating a wrapper function specific to `highlight` to not make `href_with_root_path` bigger with an extra argument?

Comment (jsha) [reply src/librustdoc/html/highlight.rs:793]: Actually I don't think you need to add `Option<HrefContext>` as a last argument. Here's the definition of `href_with_root_path`:

```
pub(crate) fn href_with_root_path(
    did: DefId,
    cx: &Context<'_>,
    root_path: Option<&str>,
) -> Result<(String, ItemType, Vec<Symbol>), HrefError> {
```

Here's the definition of `HrefContext`:

```
pub(crate) struct HrefContext<'a, 'b, 'c> {
    pub(crate) context: &'a Context<'b>,
    pub(crate) file_span: Span,
    pub(crate) root_path: &'c str,
}
```

And the code path we're talking about doesn't care about `file_span`, so all it needs are `context` and `root_path`, which are available in `href_with_root_path`.

Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:793]: I completely missed that... Sorry for being so slow to catch up and thanks a lot for taking the time to explain to me what was so obvious! Updating it then. Thanks so much for this!

Comment (jsha) [reply src/librustdoc/html/highlight.rs:793]: It was not so obvious. :-D It took me a couple rounds of reading and thinking "maybe it could work like this..." For instance my original recommendation was to hoist HrefContext to a different module, but that turned out to not be the right path.

File: src/librustdoc/html/format.rs [hunk 22]
```diff
@@ -613,7 +680,11 @@ pub(crate) fn href_with_root_path(
                     },
                 )
             } else {
-                return Err(HrefError::NotInExternalCache);
+                if matches!(def_kind, DefKind::Macro(_)) {
```
Comment (jsha) [inline src/librustdoc/html/format.rs:683]: This `if` can be joined with the `else` above it to make an `else if`.

File: src/librustdoc/html/highlight.rs (hunk 13 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs]: I made `href_with_root_path` call the function which I moved just above it. I think the end result looks quite nice. It was a great suggestion, thanks a lot for it!

**Review by jsha — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/format.rs (hunk 22 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs]: It's something I do often: keeping things "together" under a same "else" block. Not sure why. Well, fixed in any case. :)

**Review by jsha — COMMENTED** [review]
Looks good! r= me after CI passes.

**Comment by jsha** [comment]
Latest commit also looks good to me.

**Comment by GuillaumeGomez** [comment]
bors is dead? Let's retry...

@bors r=jsha

**Comment by GuillaumeGomez** [comment]
Ah, seems like there is an incident in progress: https://www.githubstatus.com/incidents/gvzcw6sd0xxb
